crypto/ct/ct_lib.c

```diff
@@ -0,0 +1,301 @@
+    return 1;
+}
+
+int SCT_set0_signature(SCT *sct, unsigned char *sig, size_t siglen)
+{
+    if (sct == NULL) {
+        CTerr(CT_F_SCT_SET0_SIGNATURE, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    if (sct->sig)
+        OPENSSL_free(sct->sig);
+    sct->sig = sig;
+    sct->siglen = siglen;
+    return 1;
+}
+
+int SCT_get_version(const SCT *sct, sct_version_t *version)
+{
+    if (sct == NULL || version == NULL) {
+        CTerr(CT_F_SCT_GET_VERSION, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *version = sct->version;
+    return 1;
+}
+
+int SCT_get_log_entry_type(SCT *sct, log_entry_type_t *entry_type)
+{
+    if (sct == NULL || entry_type == NULL) {
+        CTerr(CT_F_SCT_GET_LOG_ENTRY_TYPE, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *entry_type = sct->entry_type;
+    return 1;
+}
+
+int SCT_get0_logid(const SCT *sct, unsigned char **logid, size_t *logidlen)
+{
+    if (sct == NULL || logid == NULL || logidlen == NULL) {
+        CTerr(CT_F_SCT_GET0_LOGID, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *logid = sct->logid;
+    *logidlen = sct->logidlen;
+    return 1;
+}
+
+int SCT_get_timestamp(const SCT *sct, uint64_t *timestamp)
+{
+    if (sct == NULL || timestamp == NULL) {
+        CTerr(CT_F_SCT_GET_TIMESTAMP, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *timestamp = sct->timestamp;
+    return 1;
+}
+
+int SCT_get_signature_nid(const SCT *sct)
+{
+    if (sct == NULL) {
+        CTerr(CT_F_SCT_GET_SIGNATURE_NID, ERR_R_PASSED_NULL_PARAMETER);
+        return -1;
+    }
+    if (sct->version != SCT_V1) {
+        CTerr(CT_F_SCT_GET_SIGNATURE_NID, CT_R_SCT_UNSUPPORTED_VERSION);
+        return -1;
+    }
+    /* RFC6962 only permits two signature algorithms */
+    if (sct->hash_alg == TLSEXT_hash_sha256) {
+        if (sct->sig_alg == TLSEXT_signature_rsa)
+            return NID_sha256WithRSAEncryption;
+        if (sct->sig_alg == TLSEXT_signature_ecdsa)
+            return NID_ecdsa_with_SHA256;
+    }
+    return NID_undef;
+}
+
+int SCT_get0_extensions(const SCT *sct, unsigned char **ext, size_t *extlen)
+{
+    if (sct == NULL || ext == NULL || extlen == NULL) {
+        CTerr(CT_F_SCT_GET0_EXTENSIONS, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *ext = sct->ext;
+    *extlen = sct->extlen;
+    return 1;
+}
+
+int SCT_get0_signature(const SCT *sct, unsigned char **sig, size_t *siglen)
+{
+    if (sct == NULL || sig == NULL || siglen == NULL) {
+        CTerr(CT_F_SCT_GET0_SIGNATURE, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    *sig = sct->sig;
+    *siglen = sct->siglen;
+    return 1;
+}
+
+/* Check SCT is valid */
+int sct_check_format(const SCT *sct)
+{
+    if (sct == NULL) {
+        CTerr(CT_F_SCT_CHECK_FORMAT, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    if (sct->version == UNSET_VERSION)
+        return 0;
+    if (sct->version != SCT_V1) {
+        /* Just need cached encoding */
```

Comment:
I'd prefer sct_check_format to do what it does, i.e., check the format. If something that calls this can operate only with the cached encoding, then this should be a separate check.

Maybe remove from this MR and add back when it's actually called, so we see how it's used.